### ISSUE TITLE
Should `pl.Expr.str.split` return a struct column instead of a list column?

### ISSUE THREAD

**Issue by markfairbanks** [pola-rs/polars#2864]
Or maybe `pl.Expr.str.split` should have an `as_struct` arg that can be `True` or `False`?

This would work nicely with `.unnest()` in a situation like this.

```python
import polars as pl
from polars import col

df = pl.DataFrame(dict(x = ["a_a", "b_b", "c_c"]))

(
    df
    .with_columns(
        col('x').str.split("_")
        # .apply(tuple) # Currently needed
        .struct.rename_fields(["left", "right"])
    )
    .unnest('x')
)
```

```
┌──────┬───────┐
│ left ┆ right │
│ ---  ┆ ---   │
│ str  ┆ str   │
╞══════╪═══════╡
│ a    ┆ a     │
├╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
│ b    ┆ b     │
├╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
│ c    ┆ c     │
└──────┴───────┘
```

There might be a better way than `.apply(tuple)`, but I couldn't seem to find one. On larger datasets the `apply()` approach takes a long time to execute so I find myself using this workaround.

```python
(
    df
    .with_columns([
        pl.lit("left_right").str.split("_").alias("names"),
        col('x').str.split("_").alias("vals")
    ])
    .explode(["names", "vals"])
    .pivot(index = "x",
           columns = "names",
           values = "vals")
)
```

**Comment by ritchie46** [comment]
A `split` operation does not know how may splits it makes beforehand. In fact they may differ per row, that's why we return a list.

If you have more information than the `split` operation has beforehand (you know we return 2 fields for instance), you can easily make a struct out of it:

```python
df = pl.DataFrame(dict(x = ["a_a", "b_b", "c_c"]))

(
    df
    .with_columns(
        pl.col('x').str.split("_")
    ).with_column(
        pl.struct([
            pl.col("x").arr.get(i).alias(f"field{i}") for i in range(2)
        ]).alias("x"), 
    )
    .unnest('x')
)
```

```
shape: (3, 2)
┌────────┬────────┐
│ field0 ┆ field1 │
│ ---    ┆ ---    │
│ str    ┆ str    │
╞════════╪════════╡
│ a      ┆ a      │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
│ b      ┆ b      │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
│ c      ┆ c      │
└────────┴────────┘

```

**Comment by ghuls** [comment]
If a max_splits arg is added like suggested in https://github.com/pola-rs/polars/issues/2746, polars could know that it always has x amount of splits (if it can't be split enough, it should return None for that part).

**Comment by ritchie46** [comment]
> If a max_splits arg is added like suggested in https://github.com/pola-rs/polars/issues/2746, polars could know that it always has x amount of splits (if it can't be split enough, it should return None for that part).

That I know, but that would be a specific version of `split`. For the case in #2746 we might better create a new function `split_exact` or something similar.

For the generic `split` expression I think the `List` is a good return type.

**Comment by ghuls** [comment]
> That I know, but that would be a specific version of `split`. For the case in #2746 we might better create a new function `split_exact` or something similar.
A specific function is fine for me.

> For the generic `split` expression I think the `List` is a good return type.
I agee

**Comment by markfairbanks** [comment]
@ritchie46 - I think the `Expr.arr` namespace is exactly what I was looking for. I didn't know that existed!

Thanks for the help!

It seems like there's more of a discussion on possible other functionality now - but feel free to close this for my original request if you want.

**Comment by ritchie46** [comment]
Closed by #2872
